### PR TITLE
Alt clicking on doors sends the AI a request link to open it

### PR DESCRIPTION
**Pull request by ghost**
🆑 Kor/Thefastfoodguy
rscadd: Alt+clicking on an airlock will now send a request to the AI to open it. The door must be on camera for this to work.
/🆑

Alt+click on a door will send the AI a notice with the doors name, and the name (not real name, so disguises will work here) of the person requesting the door. Clicking the link will open the door.

There is a 10 second cooldown on each door to stop request spam.

The door must be near cameras, or the AI can not open it.

Reasoning: Opening doors is annoying busy work, it tends to generate a bunch of complaints when an AI is swamped (its not listening to me!), etc, and it'd be nice if you could let the clown into EVA and the scientist into tech storage without taking your eyes away from the murder.

### CODE REVIEW

**Comment by ghost** [comment]
I echo what tee gee said.
It being on alt click is bad.

**Comment by j-mie** [comment]
I'm ok with AltClick

**Comment by tortellinitony** [comment]
how do people have some magic super power to telepathically tell the AI to open doors. I like how you have to spam AI OPEN

**Comment by JohnGinnane** [comment]
neat idea, but what about making adding a "jump to location" thing so AIs can see who wants to go where specifically?

**Comment by Pyko1** [comment]
Maybe the door has a magic super power doorbell to AI

File: hippiestation/code/game/machinery/doors/airlock.dm [hunk 1]
```diff
@@ -1,6 +1,31 @@
...
+	var/request_cooldown = 0 //To prevent spamming requests for the AI to open
 	 
 /obj/machinery/door/airlock/clown
-	doorClose = 'sound/items/bikehorn.ogg'	
\ No newline at end of file
+	doorClose = 'sound/items/bikehorn.ogg'
+
+/obj/machinery/door/airlock/AltClick(mob/living/user)
+	if(!istype(user))
+		to_chat(user, "<span class='info'>Nice try, ghosts.</span>")
```
Comment (KayeArray) [inline hippiestation/code/game/machinery/doors/airlock.dm]: no need for this

File: hippiestation/code/game/machinery/doors/airlock.dm [hunk 2]
```diff
@@ -1,6 +1,31 @@
...
+
+	if(request_cooldown > world.time)
+		to_chat(user, "<span class='info'>The airlock's spam filter is blocking your request. Please wait at least 10 seconds between requests.</span>")
+		return
+
+	for(var/mob/living/silicon/ai/AI in GLOB.ai_list)
+		if(!AI.client)
+			continue
+		to_chat(AI, "<span class='info'><a href='?src=\ref[AI];track=[html_encode(user.name)]'><span class='name'>[user.name] ([user.GetJob()])</span></a> is requesting you to open [src]<a href='?src=\ref[AI];remotedoor=\ref[src]'>(Open)</a></span>")
+	request_cooldown = world.time + 100
```
Comment (KayeArray) [inline hippiestation/code/game/machinery/doors/airlock.dm]: magic number

File: hippiestation/code/game/machinery/doors/airlock.dm [hunk 3]
```diff
@@ -1,6 +1,31 @@
...
-	doorDeni = 'hippiestation/sound/machine/denied.ogg' 
+	doorDeni = 'hippiestation/sound/machine/denied.ogg'
+	var/request_cooldown = 0 //To prevent spamming requests for the AI to open
 	 
 /obj/machinery/door/airlock/clown
-	doorClose = 'sound/items/bikehorn.ogg'	
\ No newline at end of file
+	doorClose = 'sound/items/bikehorn.ogg'
+
+/obj/machinery/door/airlock/AltClick(mob/living/user)
```
Comment (KayeArray) [inline hippiestation/code/game/machinery/doors/airlock.dm:11]: please make it a right click option or something instead

**Review by KayeArray — CHANGES_REQUESTED** [review]

File: hippiestation/code/game/machinery/doors/airlock.dm (hunk 3 above)
Comment (ghost) [reply hippiestation/code/game/machinery/doors/airlock.dm:11]: gross

**Comment by ghost** [comment]
@JohnGinnane i made it so it tracks the user if you click on their name

**Comment by Pyko1** [comment]
time to rename ai upload to clown office and break bridge camera as a clown?

File: hippiestation/code/game/machinery/doors/airlock.dm (hunk 3 above)
Comment (KayeArray) [reply hippiestation/code/game/machinery/doors/airlock.dm:11]: alt click is gross

**Review by KayeArray — APPROVED** [review]
Looks fine, merging

**Pull request merged**